### PR TITLE
Bill product and trade pricers.

### PR DESCRIPTION
**Pull request by marc-henrard**
Pricers for Bill products and trades
- present value
- present value sensitivity
- yield and price
- currency exposure and current cash

### CODE REVIEW

File: modules/pricer/src/main/java/com/opengamma/strata/pricer/bond/DiscountingBillProductPricer.java [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+  /**
+   * Calculates the price for settlement at a given settlement date using curves.
+   * 
+   * @param bill  the bill
+   * @param provider  the discounting provider
+   * @param settlementDate  the settlement date
+   * @return the price
+   */
+  public double priceFromCurves(ResolvedBill bill, LegalEntityDiscountingProvider provider, LocalDate settlementDate) {
+    ArgChecker.isTrue(settlementDate.isBefore(bill.getNotional().getDate()),
```
Comment (jodastephen) [inline modules/pricer/src/main/java/com/opengamma/strata/pricer/bond/DiscountingBillProductPricer.java]: `ArgChecker.inOrderNotEqual`

File: modules/pricer/src/main/java/com/opengamma/strata/pricer/bond/DiscountingBillProductPricer.java [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+   * 
+   * @param bill  the bill
+   * @param provider  the discounting provider
+   * @param settlementDate  the settlement date
+   * @return the price
+   */
+  public double priceFromCurves(ResolvedBill bill, LegalEntityDiscountingProvider provider, LocalDate settlementDate) {
+    ArgChecker.isTrue(settlementDate.isBefore(bill.getNotional().getDate()), 
+        "settlement date must be before maturity date");
+    ArgChecker.isTrue(!settlementDate.isBefore(provider.getValuationDate()),
```
Comment (jodastephen) [inline modules/pricer/src/main/java/com/opengamma/strata/pricer/bond/DiscountingBillProductPricer.java]: `ArgChecker.inOrderOrEqual`

File: modules/pricer/src/main/java/com/opengamma/strata/pricer/bond/DiscountingBillProductPricer.java [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+    ArgChecker.isTrue(!settlementDate.isBefore(provider.getValuationDate()), 
+        "settlement date must be on or after valuation date");
+    IssuerCurveDiscountFactors discountFactors = provider.issuerCurveDiscountFactors(
+        bill.getLegalEntityId(), bill.getCurrency());
+    double dfMaturity = discountFactors.discountFactor(bill.getNotional().getDate());
+    RepoCurveDiscountFactors discountFactorsRepo =
+        provider.repoCurveDiscountFactors(bill.getSecurityId(), bill.getLegalEntityId(), bill.getCurrency());
+    double dfRepoSettle = discountFactorsRepo.discountFactor(settlementDate);
+    double price = dfMaturity / dfRepoSettle;
+    return price;
```
Comment (jodastephen) [inline modules/pricer/src/main/java/com/opengamma/strata/pricer/bond/DiscountingBillProductPricer.java]: Just return the calculation - no need for `price` local variable

File: modules/pricer/src/main/java/com/opengamma/strata/pricer/bond/DiscountingBillProductPricer.java [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+        "settlement date must be on or after valuation date");
+    IssuerCurveDiscountFactors discountFactors = provider.issuerCurveDiscountFactors(
+        bill.getLegalEntityId(), bill.getCurrency());
+    double dfMaturity = discountFactors.getDiscountFactors()
+        .discountFactorWithSpread(bill.getNotional().getDate(), zSpread, compoundedRateType, periodsPerYear);
+    RepoCurveDiscountFactors discountFactorsRepo =
+        provider.repoCurveDiscountFactors(bill.getSecurityId(), bill.getLegalEntityId(), bill.getCurrency());
+    double dfRepoSettle = discountFactorsRepo.discountFactor(settlementDate);
+    double price = dfMaturity / dfRepoSettle;
+    return price;
```
Comment (jodastephen) [inline modules/pricer/src/main/java/com/opengamma/strata/pricer/bond/DiscountingBillProductPricer.java]: Inline variable

File: modules/pricer/src/main/java/com/opengamma/strata/pricer/bond/DiscountingBillProductPricer.java [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+      LegalEntityDiscountingProvider provider, 
+      LocalDate settlementDate,
+      double zSpread,
+      CompoundedRateType compoundedRateType,
+      int periodsPerYear) {
+
+    ArgChecker.isTrue(settlementDate.isBefore(bill.getNotional().getDate()), 
+        "settlement date must be before maturity date");
+    ArgChecker.isTrue(!settlementDate.isBefore(provider.getValuationDate()), 
+        "settlement date must be on or after valuation date");
```
Comment (jodastephen) [inline modules/pricer/src/main/java/com/opengamma/strata/pricer/bond/DiscountingBillProductPricer.java]: ArgChecker changes

File: modules/pricer/src/main/java/com/opengamma/strata/pricer/bond/DiscountingBillTradePricer.java [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+  /**
+   * Creates an instance.
+   * 
+   * @param productPricer  the pricer for {@link ResolvedBill}
+   * @param paymentPricer  the pricer for {@link Payment}
+  */
+  public DiscountingBillTradePricer(
+      DiscountingBillProductPricer productPricer,
+      DiscountingPaymentPricer paymentPricer) {
+    this.productPricer = ArgChecker.notNull(productPricer, "productPricer");
```
Comment (jodastephen) [inline modules/pricer/src/main/java/com/opengamma/strata/pricer/bond/DiscountingBillTradePricer.java]: Blank line when parameters are on multiple lines

File: modules/pricer/src/main/java/com/opengamma/strata/pricer/bond/DiscountingBillTradePricer.java [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+  public CurrencyAmount presentValue(ResolvedBillTrade bill, LegalEntityDiscountingProvider provider) {
+    if (provider.getValuationDate().isAfter(bill.getProduct().getNotional().getDate())) {
+      return CurrencyAmount.of(bill.getProduct().getCurrency(), 0.0d);
+    }
+    CurrencyAmount pvProduct = productPricer.presentValue(bill.getProduct(), provider)
+        .multipliedBy(bill.getQuantity());
+    if (bill.getSettlement().isPresent()) {
+      DiscountFactors discountFactorsRepo = provider.repoCurveDiscountFactors(
+          bill.getProduct().getSecurityId(), bill.getProduct().getLegalEntityId(), bill.getProduct().getCurrency())
+          .getDiscountFactors();
```
Comment (jodastephen) [inline modules/pricer/src/main/java/com/opengamma/strata/pricer/bond/DiscountingBillTradePricer.java]: This kind of formatting where parameters and chained method calls are at the same indent is hard to read. Move `provider.repoCurveDiscountFactors(` to a new line.

File: modules/pricer/src/main/java/com/opengamma/strata/pricer/bond/DiscountingBillTradePricer.java [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+   * @param provider  the discounting provider
+   * @return  the present value
+   */
+  public CurrencyAmount presentValue(ResolvedBillTrade bill, LegalEntityDiscountingProvider provider) {
+    if (provider.getValuationDate().isAfter(bill.getProduct().getNotional().getDate())) {
+      return CurrencyAmount.of(bill.getProduct().getCurrency(), 0.0d);
+    }
+    CurrencyAmount pvProduct = productPricer.presentValue(bill.getProduct(), provider)
+        .multipliedBy(bill.getQuantity());
+    if (bill.getSettlement().isPresent()) {
```
Comment (jodastephen) [inline modules/pricer/src/main/java/com/opengamma/strata/pricer/bond/DiscountingBillTradePricer.java]: I think "if not then return" would be clearer logic

File: modules/pricer/src/main/java/com/opengamma/strata/pricer/bond/DiscountingBillTradePricer.java [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+        .presentValueWithZSpread(bill.getProduct(), provider, zSpread, compoundedRateType, periodsPerYear)
+        .multipliedBy(bill.getQuantity());
+    if (bill.getSettlement().isPresent()) {
+      DiscountFactors discountFactorsRepo = provider.repoCurveDiscountFactors(
+          bill.getProduct().getSecurityId(), bill.getProduct().getLegalEntityId(), bill.getProduct().getCurrency())
+          .getDiscountFactors();
+      CurrencyAmount pvSettle = paymentPricer.presentValue(bill.getSettlement().get(), discountFactorsRepo);
+      return pvProduct.plus(pvSettle);
+    }
+    return pvProduct;
```
Comment (jodastephen) [inline modules/pricer/src/main/java/com/opengamma/strata/pricer/bond/DiscountingBillTradePricer.java]: Same comments as method above

File: modules/pricer/src/main/java/com/opengamma/strata/pricer/bond/DiscountingBillTradePricer.java [hunk 10]
```diff
@@ -0,0 +1,242 @@
...
+  /**
+   * Calculates the present value of a bill trade.
+   * <p>
+   * If the settlement details are provided, the present value is the sum of the underlying product's present value
+   * multiplied by the quantity and the present value of the settlement payment if still due at the valuation date. 
+   * If not it is the underlying product's present value multiplied by the quantity.
+   * 
+   * @param bill  the bill
+   * @param provider  the discounting provider
+   * @return  the present value
```
Comment (jodastephen) [inline modules/pricer/src/main/java/com/opengamma/strata/pricer/bond/DiscountingBillTradePricer.java]: Should only be one space after `@return`

File: modules/pricer/src/main/java/com/opengamma/strata/pricer/bond/DiscountingBillTradePricer.java [hunk 11]
```diff
@@ -0,0 +1,242 @@
...
+        .multipliedBy(bill.getQuantity());
+    if (bill.getSettlement().isPresent()) {
+      DiscountFactors discountFactorsRepo = provider.repoCurveDiscountFactors(
+          bill.getProduct().getSecurityId(), bill.getProduct().getLegalEntityId(), bill.getProduct().getCurrency())
+          .getDiscountFactors();
+      PointSensitivities sensiSettle = 
+          paymentPricer.presentValueSensitivity(bill.getSettlement().get(), discountFactorsRepo).build();
+      return sensiProduct.combinedWith(sensiSettle);
+    }
+    return sensiProduct;
```
Comment (jodastephen) [inline modules/pricer/src/main/java/com/opengamma/strata/pricer/bond/DiscountingBillTradePricer.java]: Same comments as method above

File: modules/pricer/src/main/java/com/opengamma/strata/pricer/bond/DiscountingBillTradePricer.java [hunk 12]
```diff
@@ -0,0 +1,242 @@
...
```
Comment (jodastephen) [inline modules/pricer/src/main/java/com/opengamma/strata/pricer/bond/DiscountingBillTradePricer.java]: Same comments as method above

File: modules/pricer/src/main/java/com/opengamma/strata/pricer/bond/DiscountingBillTradePricer.java [hunk 13]
```diff
@@ -0,0 +1,242 @@
...
+   * @param compoundedRateType  the compounded rate type
+   * @param periodsPerYear  the number of periods per year
+   * @return the currency exposure
+   */
+  public MultiCurrencyAmount currencyExposureWithZSpread(
+      ResolvedBillTrade bill, 
+      LegalEntityDiscountingProvider provider,
+      double zSpread,
+      CompoundedRateType compoundedRateType,
+      int periodsPerYear) {
```
Comment (jodastephen) [inline modules/pricer/src/main/java/com/opengamma/strata/pricer/bond/DiscountingBillTradePricer.java]: Blank line here

File: modules/pricer/src/main/java/com/opengamma/strata/pricer/bond/DiscountingBillTradePricer.java [hunk 14]
```diff
@@ -0,0 +1,242 @@
...
+  }
+
+  /**
+   * Calculates the present value of a bill trade.
+   * <p>
+   * If the settlement details are provided, the present value is the sum of the underlying product's present value
+   * multiplied by the quantity and the present value of the settlement payment if still due at the valuation date. 
+   * If not it is the underlying product's present value multiplied by the quantity.
+   * 
+   * @param bill  the bill
```
Comment (jodastephen) [inline modules/pricer/src/main/java/com/opengamma/strata/pricer/bond/DiscountingBillTradePricer.java]: In all methods on the class, this parameter should be called `trade`

File: modules/pricer/src/test/java/com/opengamma/strata/pricer/bond/DiscountingBillProductPricerTest.java [hunk 15]
```diff
@@ -0,0 +1,260 @@
...
+
+  private static final ReferenceData REF_DATA = ReferenceData.standard();
+  private static final LocalDate VAL_DATE = date(2018, 6, 20);
+
+  // Bill
+  private static final StandardId SECURITY_ID = StandardId.of("OG-Ticker", "GOVT1-BOND1");
+  private static final StandardId ISSUER_ID = StandardId.of("OG-Ticker", "GOVT1");
+  private static final BillYieldConvention YIELD_CONVENTION = BillYieldConvention.INTEREST_AT_MATURITY;
+  
+
```
Comment (jodastephen) [inline modules/pricer/src/test/java/com/opengamma/strata/pricer/bond/DiscountingBillProductPricerTest.java]: Only one blank line

File: modules/pricer/src/test/java/com/opengamma/strata/pricer/bond/DiscountingBillTradePricerTest.java [hunk 16]
```diff
@@ -0,0 +1,323 @@
...
+public class DiscountingBillTradePricerTest {
+
+  private static final ReferenceData REF_DATA = ReferenceData.standard();
+  private static final LocalDate VAL_DATE = date(2018, 6, 20);
+
+  // Bill
+  private static final StandardId SECURITY_ID = StandardId.of("OG-Ticker", "GOVT1-BOND1");
+  private static final StandardId ISSUER_ID = StandardId.of("OG-Ticker", "GOVT1");
+  private static final BillYieldConvention YIELD_CONVENTION = BillYieldConvention.INTEREST_AT_MATURITY;
+
```
Comment (jodastephen) [inline modules/pricer/src/test/java/com/opengamma/strata/pricer/bond/DiscountingBillTradePricerTest.java]: Only one blank line

Comment (yukiiwashita) [reply modules/pricer/src/test/java/com/opengamma/strata/pricer/bond/DiscountingBillTradePricerTest.java]: A blank line removed.

**Pull request merged**